[css-grid] Anecstor subgrids' gutters should add to the extra layer of margin for descendant subgrids

sammygill opened the pull request:
#### dedaa29e2a66866ec206ffa113d0953eabc092ad
<pre>
[css-grid] Anecstor subgrids&apos; gutters should add to the extra layer of margin for descendant subgrids
<a href="https://bugs.webkit.org/show_bug.cgi?id=260532">https://bugs.webkit.org/show_bug.cgi?id=260532</a>
rdar://114271857

Reviewed by Matt Woodrow.

268947@main starts to accumulate the margin, border, and padding of
nested subgrids to act as a single layer of additional margin for the
track. This was needed for the specific scenario in which there were
subgrids without any items as for other types of grid items we would
walk back up the ancestor chain to compute this value. Subgrids are
treated differently so that behavior is not done and we needed this
extra change.

This patch aims to expand on the scenario by including the gutter sizes
from ancestor subgrids to this extra layer of margin. We can do this
by walking back up the ancestor chain to compute the total value of the
gutters that would act as an additional margin for items in the track.

This computation is only done if the subgrid&apos;s start/end position is one
of the lines of the track in the outer grid. Otherwise, this subgrid
spans *through* the track and cannot have the any gutters from subgrids
act as margins for this track.

Notably, we want to make sure that we do not add this value directly
to currentAccumulatedMbp and recurse into
accumulateIntrinsicSizesForTrack with it. We will compute this value
for each subgrid that acts as a grid item by walking up the ancestor
chain. This is because it is difficult to know ahead of time whether
a subgrid that is positioned in a track will have nested subgrids that
are also positioned in the same track and avoid adding the gutters
multiple times.

* LayoutTests/TestExpectations:
* Source/WebCore/rendering/GridLayoutFunctions.cpp:
(WebCore::GridLayoutFunctions::hasRelativeOrIntrinsicSizeForChild):
* Source/WebCore/rendering/GridLayoutFunctions.h:
* Source/WebCore/rendering/GridTrackSizingAlgorithm.cpp:
(WebCore::GridTrackSizingAlgorithm::accumulateIntrinsicSizesForTrack):

Canonical link: <a href="https://commits.webkit.org/269194@main">https://commits.webkit.org/269194@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/3f4377a8ef916fae7d6450d512dc6535d2636b7f

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/21816 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/22042 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/22865 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/23688 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/20195 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/22059 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/26272 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/22341 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/21309 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/22044 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/21629 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/18893 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/24540 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/18798 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/19750 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/26033 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/19827 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/19963 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/23898 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/20429 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/17401 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/19772 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/5214 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/23982 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/20368 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

mattwoodrow commented on Source/WebCore/rendering/GridTrackSizingAlgorithm.cpp:
```diff
@@ -1389,8 +1389,38 @@ void GridTrackSizingAlgorithm::accumulateIntrinsicSizesForTrack(GridTrack& track
             GridSpan span = m_renderGrid->gridSpanForChild(*gridItem, m_direction);
 
+            auto extraMarginFromSubgridAncestorGutters = [&]() -> std::optional<LayoutUnit> {
+                if (span.startLine() != trackIndex && span.endLine() - 1 != trackIndex)
```
Do we not need to add the gutter padding for tracks that are in the middle of the sub gridded span?

sammygill replied:
Based off my current reading of the spec I don't *think* so:

> Meanwhile, half the size of the difference between the [subgrid](https://drafts.csswg.org/css-grid-2/#subgrid)’s [gutters](https://drafts.csswg.org/css-align-3/#gutter) ([row-gap](https://drafts.csswg.org/css-align-3/#propdef-row-gap)/[column-gap](https://drafts.csswg.org/css-align-3/#propdef-column-gap)) and its [parent grid](https://drafts.csswg.org/css-grid-2/#parent-grid)’s gutters is applied as an extra layer of (potentially negative) margin to the items not at those edges.

Since we're adding extra margins in particular to the subgrids, I think we only want to affect the tracks where the edges are

mattwoodrow commented on Source/WebCore/rendering/GridTrackSizingAlgorithm.cpp:
```diff
@@ -1389,8 +1389,38 @@ void GridTrackSizingAlgorithm::accumulateIntrinsicSizesForTrack(GridTrack& track
+                const auto* currentAncestor = dynamicDowncast<RenderGrid>(gridItem->parent());
+                auto flowAwareDirection = iterator.direction();
+
+                while (currentAncestor && currentAncestor->isSubgrid(flowAwareDirection)) {
```
I wouldn't block on this, but it seems that a lot of the complexity here is iterating up the sub grid ancestors and converting coordinate space.

It would be somewhat nice if we had a `forEachSubgridAncestor` iterator thing that we could use in all the places we traverse the sub grid ancestors, which put all of that code in one spot.

sammygill replied:
Yeah I agree. We probably do this in a few other places and should definitely have something. I'll try to see if I can come up with such a thing and try to insert it where it is needed in the current code

sammygill committed:
[css-grid] Anecstor subgrids' gutters should add to the extra layer of margin for descendant subgrids
https://bugs.webkit.org/show_bug.cgi?id=260532
rdar://114271857

Reviewed by Matt Woodrow.

268947@main starts to accumulate the margin, border, and padding of
nested subgrids to act as a single layer of additional margin for the
track. This was needed for the specific scenario in which there were
subgrids without any items as for other types of grid items we would
walk back up the ancestor chain to compute this value. Subgrids are
treated differently so that behavior is not done and we needed this
extra change.

This patch aims to expand on the scenario by including the gutter sizes
from ancestor subgrids to this extra layer of margin. We can do this
by walking back up the ancestor chain to compute the total value of the
gutters that would act as an additional margin for items in the track.

This computation is only done if the subgrid's start/end position is one
of the lines of the track in the outer grid. Otherwise, this subgrid
spans *through* the track and cannot have the any gutters from subgrids
act as margins for this track.

Notably, we want to make sure that we do not add this value directly
to currentAccumulatedMbp and recurse into
accumulateIntrinsicSizesForTrack with it. We will compute this value
for each subgrid that acts as a grid item by walking up the ancestor
chain. This is because it is difficult to know ahead of time whether
a subgrid that is positioned in a track will have nested subgrids that
are also positioned in the same track and avoid adding the gutters
multiple times.

* LayoutTests/TestExpectations:
* Source/WebCore/rendering/GridLayoutFunctions.cpp:
(WebCore::GridLayoutFunctions::hasRelativeOrIntrinsicSizeForChild):
* Source/WebCore/rendering/GridLayoutFunctions.h:
* Source/WebCore/rendering/GridTrackSizingAlgorithm.cpp:
(WebCore::GridTrackSizingAlgorithm::accumulateIntrinsicSizesForTrack):

Canonical link: https://commits.webkit.org/269194@main
